### ISSUE TITLE
Realistic player defense overhaul

### ISSUE THREAD

**Issue by bananaedmonkey** [CleverRaven/Cataclysm-DDA#25064]
Presently in the cataclysm, even a god with a melee skill of 20 will haplessly watch a zombie bite the 10% of his arm that is uncovered rather than the part that is armored. In reality, a skilled fighter will maneuver his bodyparts so that the uncovered portion is never in harms way, increasing the effective coverage of a elbow pad that only literally covers 30% of his arm.

Additionally, only characters using a martial art can comprehend the concept of blocking with their arms or legs right now. I think this should be changed so that it is a innate ability to all characters at all times that scales as their melee skill increases.

To solve the issues above I present a solution that consists of several parts:

Part 1: Solving the issue of characters not maneuvering to avoid blows to undefended parts of arms and legs.

I believe that certain items should be given the DYNAMIC_DEFENSE flag. These items have their coverage scale with the user's defense rating. An equation similar to 

effective coverage = item.coverage * (( 1 + melee_skill) / 2)

should work well, but that is open to discussion.

For balancing reasons, some items given the DYNAMIC_DEFENSE flag should have their literal coverage reduced.

~~Part 2: Solving the issue of characters being able to use active defense in all directions~~

This has been shown to be a misstep in game design, so part 2 can be ignored.

Part 3: Solving the issue of only black belt masters understanding the concept of "if knife com 2 throat, move arm in wae"

Part 3 will make any character attempt to block a blow to the head or torso with a arm or leg. 

If part 3 is implemented, this whole combat change would really come together. A blow to the head would be redirected to an arm, which in turn would be redirected to an armored part of said arm. Just like how a skilled fighter would operate in real life.

Part 4: Smoothing out the blocking curve
As pointed out by @kevingranade a characters in this game go form inept at melee combat to a dodging and blocking god once that critical mass is reached.

To reduce the sheer cliff that is combat scaling blocking should be split up into two parts. One part is a flat redirection that varies by +- (melee_skill + 2). The damage redirected will be applied to a character's arms instead.

The other part of blocking with be the chance of a complete redirection of damage. It will be a percentage chance based off whatever it is presently.

Parts 1 should also function fine without part 3, but the change would not be as revolutionary 

Effects of implementing all the parts:

1. More varied playstyles would be easier to pull off. A lightly armored survivor built around dodging and deflecting blows to heavily armored sections of their limbs could finally compete with a survivor brainlessly decked out in power armor. This would allow mutation trees like the bird tree to shine since their weight limit would not have to be wasted on heavy armor.

2. A far more realistic and dynamic approach to melee defense would be accomplished

An addition that could be built on top of part 1 and part 3:
3. CBMS:
   a. HyperReaction CBM: Allows dynamic defense to effect bullets and other ranged attacks (Wonder Woman anyone?)
    

Expect a feeble attempted to implement part 1 in the next week or so since I seem to like biting off more than I could eat in a year.

**Comment by woutershep** [comment]
> Part 3 will make any character attempt to block a blow to the head or torso with a arm or leg.

I don't know about you but if i tried to block a blow to my head with my leg in a close combat situation id probably fall over. I think if you want to pursue leg blocking for novices that it needs a chance to get effect_downed that decreases the more skilled you are. I could even imagine going the extra mile and making you better at blocking with your legs once you have effect_downed, since you know if you're laying on your back you don't need them for balance anymore and can wiggle and kick with them to the full extent possible.

I'm personally also not quite sure about the concept of tying dynamic defense to movement. I can imagine one taking a step back or a sidestep around them to maneuver ones self a bit better but that they keep their eyes on the enemies infront of them. It would be really stupid if you turn your back to them in that scenario. As an alternative i would propose locking it in the direction of the zombie that attacked you first after having moved so that you will gain tactical disadvantage still when you're pinned between two zombies without the strange scenario where stepping backwards in a door to funnel the group infront of you into a chokepoint would lead to having your back full of clawmarks.

To be clear i'm imaging the following scenario:

<samp>........
.XZX.... (with Z being an X)
.O@O..z.
.OOO....</samp>

At which point tactically you want the one zombie to be
in the way of the other zombie so he doesnt flank you.
So you step North West.

Your defense rules will lead to

<samp>XXO.....
X@Z..... (with the Z being an O)
OOO...z.
........</samp>

And i lose the ability to defend. That seems like the dumbest thing to naturally do.
I would expect anyone that is not a vegetable to have the following outcome:

<samp>OOX.....
O@Z..... (with the Z being an X)
OOX...z.
........</samp>

It would be a simple sidestep after all.

**Comment by kevingranade** [comment]
Tl;dr I'm interested in something like parts 1 and 3 being implemented.  I am very against introducing facing into melee combat.

The game has no good way of displaying or updating facing, it would require very invasive changes to the way combat and movement work, and the benefits are small.

As for points 1 and 3, the main problem with blocking and dodging IMO is that they're all-or-nothing, this is what leads to the inability of average characters to block and dodge, it is because giving them significant ability suddenly enables them to negate many sources of damage entirely.  Because of this, blocking ability is reserved for experts.

What I want to happen is an adjustment to both blocking and dodging to mitigate damage from attacks, at low levels it will mitigate a little damage and have a chance of redirecting it, at higher levels it would consistently negate damage, and there would be a smooth transition in between instead of a sudden jump from "cant' block consistently, no effect" to "blocks consistently, immune to damage".

**Comment by bananaedmonkey** [comment]
I have updated the issue to address the points that @kevingranade has brought up.

I while I only touched on re-balancing blocking, I was thinking that dodge could be re-balanced so that not all dodges give 100% damage resistance. There could be 4 tiers of dodges such that

tier 1 negates 25% of damage
tier 2 negates 50% of damage
tier 3 negates 75% of damage
tier 4 negates 100% of damage

Blocking and dodge could be made very distinct since Blocking would be based around a very reliable flat rate and dodging would be less reliable and based off percentages.

But that should probably be in a separate issue, and this issue is already all over the place as is.

Another item I wanted to touch upon: there are a lot of items like "bone arm-guards" that already have a high coverage. (Bone arm-guards have a coverage of 95). Should that number be nerfed  or should I change the scaling of effective coverage to be more like 

effective coverage = coverage + (coverage * ((melee_skill) / 20). 

I would greatly appreciate community input here.
